### DISCUSSION TITLE
How to show iterations per second (like tdqm)

### DISCUSSION THREAD

**Discussion by brando90** [WoLpH/python-progressbar#253] [Q&A]
Original discussion: https://github.com/WoLpH/python-progressbar/issues/250

e.g. tqdm has it:
```
 21%|██████████████████████████████████████████████████████████████████▉            21/100 [01:45<04:27,  3.39s/it]
```
----
Other message:

I've actually had a similar question before and added a widget for it: #98

But...

I think you are looking for the AdaptiveTransferSpeed: https://progressbar-2.readthedocs.io/en/latest/progressbar.widgets.html#progressbar.widgets.AdaptiveTransferSpeed
Or the FileTransferSpeed: https://progressbar-2.readthedocs.io/en/latest/progressbar.widgets.html#progressbar.widgets.FileTransferSpeed

Instead of using B as the unit parameter, you want it in this case.
Note that by default it uses the si prefixes.
Oddly enough I remembering implementing this in a separate widget but I can't find it anymore... #98

----
https://stackoverflow.com/questions/30834730/how-to-print-iterations-per-second
https://github.com/tqdm/tqdm/discussions/1211

**Comment by wolph** [comment]
```python
import time
import progressbar


widgets = [
    progressbar.Percentage(),
    progressbar.Bar(),
    ' ', progressbar.SimpleProgress(),
    ' ', progressbar.ETA(),
    ' ', progressbar.AdaptiveTransferSpeed(unit='it'),
]

bar = progressbar.ProgressBar(widgets=widgets)
for i in bar(range(100)):
    time.sleep(0.2)
    bar.update(i)
```

This does roughly what you want I think :)
```
 19%|##########                                           | 19 of 100 ETA:   0:00:17   4.9 it/s
```

**Reply by brando90** [reply]
@WoLpH apologies I should have been clear. What I really want is the standard progress bar already does but add it/s

```
 99% (9998 of 10000) |########## | Elapsed Time: 1 day, 16:35:09 ETA:   0:00:26
```

will go through your suggested solution now though.

**Reply by brando90** [reply]
btw, WoLpH, I am confused about something. In the past whenever I have made a progressbar **without** a context manager it always seems that the progressbar thinks it's not done when my loop is actually done (but it seems I do call update enough times from the behaviour I'd expect)

```
import time
import progressbar

bar = progressbar.ProgressBar(max_value=5)
for i in range(5):
    time.sleep(1)
    bar.update(i)
```

why does my code above not finish the progress bar despite the max values matching, similar to what you did?
```
import time
import progressbar
bar = progressbar.ProgressBar(max_value=10)
for i in range(10):
    time.sleep(0.1)
    time.sleep(1)
    bar.update(i)
 90% (9 of 10) |######################   | Elapsed Time: 0:00:09 ETA:   0:00:01
````

**Reply by wolph** [reply]
> @WoLpH apologies I should have been clear. What I really want is the standard progress bar already does but add it/s

That's also an option. In that case your options are:

1. inheriting the default progressbar and overriding the `default_widgets()` method
2. Setting the widgets manually and still using the `default_widgets()` method. Here's an example:

```python
import time
import progressbar

bar = progressbar.ProgressBar()
bar.widgets = bar.default_widgets() + [
    ' ', progressbar.AdaptiveTransferSpeed(unit='it'),
]

for i in bar(range(100)):
    time.sleep(0.2)
    bar.update(i)
```

**Reply by wolph** [reply]
> btw, WoLpH, I am confused about something. In the past whenever I have made a progressbar **without** a context manager it always seems that the progressbar thinks it's not done when my loop is actually done (but it seems I do call update enough times from the behaviour I'd expect)

You are setting `max_value` to 10, but `range(10)` only goes up to 9 :)

Additionally, in some cases it can also be caused by a performance optimization. If you are calling update for every byte while downloading a 1 gigabyte file, that would mean 1 billion calls to the progressbar. The progressbar automatically tries to update only when needed which is  usually a good thing.

In some cases however... the progressbar does not know when you are finished, to fix that you can either use the contextmanager to tell it, or you can manually call `bar.finish()` when you are done.

**Reply by brando90** [reply]
> > btw, WoLpH, I am confused about something. In the past whenever I have made a progressbar **without** a context manager it always seems that the progressbar thinks it's not done when my loop is actually done (but it seems I do call update enough times from the behaviour I'd expect)
> 
> You are setting `max_value` to 10, but `range(10)` only goes up to 9 :)
> 
> Additionally, in some cases it can also be caused by a performance optimization. If you are calling update for every byte while downloading a 1 gigabyte file, that would mean 1 billion calls to the progressbar. The progressbar automatically tries to update only when needed which is usually a good thing.
> 
> In some cases however... the progressbar does not know when you are finished, to fix that you can either use the contextmanager to tell it, or you can manually call `bar.finish()` when you are done.

oh, ok. I thought that 0-9 there are 10 elements so 10 updates. But the progressbar exects n+1 updates to know it's done it seems. I assume the example you sent and the context manager know by the end of the loop/or end of context.

Thanks!

**Reply by brando90** [reply]
@WoLpH I think I found what I want:
```
import time
import progressbar
widgets = [
    progressbar.Percentage(),
    ' ', progressbar.SimpleProgress(format=f'({progressbar.SimpleProgress.DEFAULT_FORMAT})'),
    ' ', progressbar.Bar(),
    ' ', progressbar.Timer(), ' | '
                              ' ', progressbar.ETA(), ' | ',
    ' ', progressbar.AdaptiveTransferSpeed(unit='it'),
]
bar = progressbar.ProgressBar(widgets=widgets)
for i in bar(range(100)):
    time.sleep(0.1)
    bar.update(i)
100% (100 of 100) |#####| Elapsed Time: 0:00:10 |  Time:  0:00:10 |    9.8 it/s

```

the only thing that makes me sad is that the bar is so small :( how do I make it bigger. Everything else is more or less where I'd want it.

**Reply by wolph** [reply]
You can choose to make things less verbose. The `Timer` for example: https://progressbar-2.readthedocs.io/en/latest/progressbar.widgets.html#progressbar.widgets.Timer
If, instead of `Elapsted Time: %(elapsed)s` you use something like this: `e %(elapsed)s`

You can make it as small or wide as you want. Perhaps you don't need the `SimpleProgress` either if you already use a `Bar`.

**Reply by wolph** [reply]
> oh, ok. I thought that 0-9 there are 10 elements so 10 updates. But the progressbar exects n+1 updates to know it's done it seems. I assume the example you sent and the context manager know by the end of the loop/or end of context.

The `max_value` specifies the actual value, you could even have floating point numbers such as `2.5` if you wanted to.
